### PR TITLE
Work on fixing #1838

### LINKED ISSUE #849: Generic types in wildcard bounds
Generic types in wildcard bounds are not checked during common assignment check. For example, neither of the expected errors are issued below:

```
class Gen<G> { }
void polyAll(Gen<Gen<@H1S2 Object>> genGenNonNull) {
    //:: error: (assignment.type.incompatible)
    Gen<@H1Top ? extends @H1Top Gen<@H1Top Object>> a = genGenNonNull;
}

void interning(Gen<Gen<@Interned Object>> genGenNonNull) {
    //:: error: (assignment.type.incompatible)
    Gen<@UnknownInterned ? extends @UnknownInterned Gen<@UnknownInterned Object>> a = genGenNonNull;
}
```

However, the Nullness Checker does correctly issue the error:

```
 void nullness(Gen<Gen<@NonNull Object>> genGenNonNull) {
    //:: error: (assignment.type.incompatible)
    Gen<@Nullable ? extends @Nullable Gen<@Nullable Object>> a = genGenNonNull;
}
```

### PR DESCRIPTION
**Pull request by wmdietl**
Fix #849 and #1838 by correctly handling generics in subtype tests.
Refactors interactions between the type hierarchies and the visitor history.

### CODE REVIEW

**Comment by wmdietl** [comment]
@smillst Could you have a look at this? Some tests now fail with the KeyFor hierarchy. I think many of those were masked previously, because as soon as one hierarchy was OK, all following hierarchies were ignored. I think the tests that I skipped are all caused by problems with KeyFor.
Maybe there is something about the KeyFor hierarchy or defaults that needs adapting?
We can go through the other changes in the meeting tomorrow.

File: checker/tests/nullness/generics/MyMap.java [hunk 1]
```diff
@@ -1,6 +1,8 @@
 import java.util.Map;
 import org.checkerframework.checker.nullness.qual.Nullable;
 
+// @skip-test issue with @KeyFor
```
Comment (smillst) [inline checker/tests/nullness/generics/MyMap.java]: checker/tests/nullness/generics/MyMap.java:16: error: [enhancedfor.type.incompatible] incompatible types in enhanced for loop.
```
  found   : @UnknownKeyFor Entry<@KeyFor("map") ?[ extends K[ extends @KeyFor("map") Object super @KeyFor("map") Void] super @KeyFor("map") Void], ?[ extends V[ extends @UnknownKeyFor Object super @KeyForBottom Void] super @KeyForBottom Void]>
  required: @UnknownKeyFor Entry<?[ extends K[ extends @UnknownKeyFor Object super @KeyForBottom Void] super @KeyForBottom Void], ?[ extends V[ extends @UnknownKeyFor Object super @KeyForBottom Void] super @KeyForBottom Void]>
```

File: framework/tests/all-systems/Issue1587b.java [hunk 2]
```diff
@@ -25,6 +25,7 @@
         Iterable<R> q();
     }
 
+    @SuppressWarnings("keyfor:type.argument.type.incompatible")
```
Comment (smillst) [inline framework/tests/all-systems/Issue1587b.java]: This is only only remaining KeyFor regression.

**Comment by smillst** [comment]
> Could you have a look at this? Some tests now fail with the KeyFor hierarchy. I think many of those were masked previously, because as soon as one hierarchy was OK, all following hierarchies were ignored.

The KeyFor Checker is a separate checker, so the failures weren't masked because of this.  They were masked because of KeyFor's DefaultTypeHierarchy overrides isSubtype(ATM, ATM).  I added the code to isSubtype(..., AnnotationMirror top) and that fixes all but one of the regressions.

We still need to figure out that last regression though.

File: framework/tests/all-systems/Issue1587b.java [hunk 3]
```diff
@@ -25,6 +25,7 @@
         Iterable<R> q();
     }
 
+    @SuppressWarnings("type.argument.type.incompatible") // TODO: FIX
```
Comment (smillst) [inline framework/tests/all-systems/Issue1587b.java]: This is the only remaining regression.

File: framework/src/org/checkerframework/framework/type/DefaultRawnessComparer.java [hunk 4]
```diff
@@ -253,50 +269,55 @@ public Boolean visitWildcardSupertype(
...
-        return false;
+        // Needed to prevent infinite recursion e.g. in
+        // tests/all-systems/java8inference/Issue1815.java
+        visited.add(subtype, supertype, currentTop, true);
+
+        // TODO: tests/all-systems/InferTypeArgsCondtionalExpression.java
+        // breaks with this:
+        //Boolean result = this.visit(subtypeUb, supertype, visited);
+        // it doesn't make sense to me to compare subtype and subtypeUB
+        Boolean result = this.visit(subtype, subtypeUb, visited);
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/DefaultRawnessComparer.java]: Highlighting this comment and the test that fails when using a more meaningful check.

File: checker/tests/nullness/java8/Issue579.java [hunk 5]
```diff
@@ -1,6 +1,8 @@
 // Test case for Issue579
 // https://github.com/typetools/checker-framework/issues/579
 
+// @skip-test caused by DefaultRawnessComparer.visitNull_Declared
```
Comment (wmdietl) [inline checker/tests/nullness/java8/Issue579.java]: Note the strange failure when using a meaningful implementation of DefaultRawnessComparer.visitNull_Declared.

**Comment by wmdietl** [comment]
Thanks for your cleanups! I've done a few more things and highlighted two strange things in comments.

**Comment by smillst** [comment]
I fixed the two things you commented one.  We still have the framework/tests/all-systems/Issue1587b.java regression to figure out.

File: framework/src/main/java/org/checkerframework/framework/type/DefaultTypeHierarchy.java [hunk 6]
```diff
@@ -1143,7 +1096,13 @@ protected boolean visitWildcardSubtype(
                 final AnnotatedTypeMirror sourceTypeArg = asSuperAdt.getTypeArguments().get(0);
                 final AnnotatedTypeMirror resultTypeArg = resultAtd.getTypeArguments().get(0);
                 resultTypeArg.clearAnnotations();
-                resultTypeArg.addAnnotations(sourceTypeArg.getAnnotations());
+                if (resultTypeArg.getKind() == TypeKind.TYPEVAR) {
```
[the remark this answers is not in the document]
Comment (smillst) [reply framework/src/main/java/org/checkerframework/framework/type/DefaultTypeHierarchy.java]: This makes sense to me, too.  (Although, I don't really understand why Enums are special cased here in the first place.

**Pull request merged**